diff --git a/host/libraries/libbladeRF/src/driver/thirdparty/adi/util.c b/host/libraries/libbladeRF/src/driver/thirdparty/adi/util.c
index a50f8b85..d1fb3bd6 100644
--- a/ad9361/sw/util.c
+++ b/ad9361/sw/util.c
@@ -331,7 +331,7 @@ uint32_t find_first_bit(uint32_t word)
 *******************************************************************************/
 void * ERR_PTR(long error)
 {
-	uint64_t val = error;
+	uintptr_t val = error;
 	return (void *) val;
 }
