diff --git no_os_axi_io.h no_os_axi_io.h
index c346559..03311d0 100644
--- no_os_axi_io.h
+++ no_os_axi_io.h
@@ -40 +40 @@
-int32_t no_os_axi_io_read(uint32_t base, uint32_t offset, uint32_t *data);
+int32_t no_os_axi_io_read(void *ctx, uint32_t base, uint32_t offset, uint32_t *data);
@@ -43 +43 @@ int32_t no_os_axi_io_read(uint32_t base, uint32_t offset, uint32_t *data);
-int32_t no_os_axi_io_write(uint32_t base, uint32_t offset, uint32_t data);
+int32_t no_os_axi_io_write(void *ctx, uint32_t base, uint32_t offset, uint32_t data);
